lib/cosmos/streams/tcpip_client_stream.rb

```diff
@@ -88,17 +88,23 @@ def connect_nonblock(socket, addr)
       rescue IO::WaitWritable
         begin
           _, sockets, _ = IO.select(nil, [socket], nil, @connect_timeout) # wait 3-way handshake completion
-        rescue Errno::ENOTSOCK
+        rescue IOError, Errno::ENOTSOCK
           raise "Connect canceled"
         end
         if sockets and !sockets.empty?
           begin
             socket.connect_nonblock(addr) # check connection failure
-          rescue Errno::EISCONN
+          rescue IOError, Errno::ENOTSOCK
+            raise "Connect canceled"
+          rescue Errno::EINPROGRESS
+            retry
+          rescue Errno::EISCONN, Errno::EALREADY
           end
         else
           raise "Connect timeout"
         end
+      rescue IOError, Errno::ENOTSOCK
+        raise "Connect canceled"
       end
```

Comment:
This is ugly but it handles lots more of the rarer cases when using connect_nonblock